### PR TITLE
source-mysql: Throughput Optimization

### PR DESCRIPTION
**Pull request by willdonnelly**
**Description:**

Part of https://github.com/estuary/connectors/issues/2885

The third in the series of SQL CDC optimization PRs along with [PostgreSQL](https://github.com/estuary/connectors/pull/2990) and [SQL Server](https://github.com/estuary/connectors/pull/3063), this one for MySQL.

- Introducing automated throughput benchmarks
- Implementing MySQL specific implementations of the commit event and change event interfaces
- Directly assembling the JSON representing a captured document instead of using reflection on a Go map.
- Offloading as much of the change event processing as possible into shared metadata which can be reused across multiple changes.
    - Note that there's probably still a little bit of juice left to squeeze out here by caching replication metadata across multiple RowsEvents for the same table, but that would require some cache invalidation logic which would be tricky to get exactly right and it doesn't appear to be necessary just yet.
- Optimizing change event processing by using datatype-specific transcoders which go directly from MySQL backfill/replication values to bytes appended to the serialized JSON output.

In my tests these changes all together roughly doubled the throughput from the current MySQL baseline (which, as with SQL Server, was already increased by about 50% from the improvements to the shared sqlcapture code). Current throughput numbers in my benchmarks are something like:

    BenchmarkBackfill_SmallRows_5M-2              1       279.2 MBps      1122755 rows/sec          4.45 seconds
    BenchmarkBackfill_ManyTypes_5M-2              1       149.5 MBps       258990 rows/sec         19.31 seconds
    BenchmarkBackfill_LargeText_1M-2              1       475.5 MBps       129046 rows/sec          7.74 seconds
    BenchmarkReplication_SmallRows_5M-2           1       269.4 MBps       844374 rows/sec          5.92 seconds
    BenchmarkReplication_ManyTypes_5M-2           1       158.4 MBps       244758 rows/sec         20.43 seconds
    BenchmarkReplication_LargeText_1M-2           1       651.2 MBps       173426 rows/sec          5.76 seconds

It's probably possible to squeeze out a bit more performance on the `ManyTypes` test case, timestamps and dates are as relatively slow here as they are in the other databases, but otherwise I think those are pretty good numbers.

Test snapshots are unaltered, however I did add a new datatype test case for null-valued JSON columns.

**Workflow steps:**

None.

### CODE REVIEW

File: source-mysql/datatypes.go [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+
+type mysqlTextTranscoderReplication struct {
+	Charset string
+}
+
+func (t *mysqlTextTranscoderReplication) TranscodeJSON(buf []byte, val any) ([]byte, error) {
+	if str, ok := val.(string); ok {
+		return json.AppendEscape(buf, str, 0), nil
+	} else if bs, ok := val.([]byte); ok {
+		// TODO(wgd): Figure out how exactly to optimize this. The UTF-8 case could
```
Comment (williamhbaker) [inline source-mysql/datatypes.go:516]: Just noting this TODO, which is probably not something intended to be addressed in this PR, but rather a later thing I imagine.

File: source-mysql/replication.go [hunk 2]
```diff
@@ -523,203 +504,234 @@ func (rs *mysqlReplicationStream) handleRowsEvent(ctx context.Context, event *re
 		return fmt.Errorf("unknown key columns for stream %q", streamID)
 	}
 
-	var nonTransactionalTable = rs.isNonTransactional(streamID)
+	// Construct efficient indices and transcoder arrays for value processing
+	//
+	// TODO(wgd): Hoist this to a higher level
```
Comment (williamhbaker) [inline source-mysql/replication.go:509]: Also noting this TODO, which seems like it may be more relevant to address now, although I'm not 100% sure if that was the intention.

**Review by williamhbaker — APPROVED** [review]
LGTM
